drivers/volume/portworx/portworx.go

```diff
@@ -510,174 +514,1520 @@ func (d *portworx) RecoverDriver(n node.Node) error {
+							}
+						}
+
+						// Check node value is less than spec value
+						if itval >= mValue {
+							logrus.Infof("ReplicaAntiAffinity with matchExpression Operator:'%v', is placed on node:%v having node value(%v) greater than Spec value (%v)  for key:(%v:%v)", matchExp.Operator, itval, mValue, node.Id, matchExp.Key, tval)
+							return true, nil
+						}
+						return false, &ErrFailedToInspectVolume{
+							Cause: fmt.Sprintf("ReplicaAntiAffinity with matchExpression Operator:'%v', is placed on node:%v having node value(%v)  less than Spec value (%v)  for key:(%v:%v)", matchExp.Operator, itval, mValue, node.Id, matchExp.Key, tval),
+						}
+
+					default:
+						return false, &ErrFailedToInspectVolume{
+							Cause: fmt.Sprintf("ReplicaAntiAffinity with UNKNOWN matchExpression Operator:'%v' is placed on node:%v having label key:(%v:%v)", matchExp.Operator, node.Id, matchExp.Key, tval),
+						}
+					}
+
+				} else {
+					logrus.Infof("ReplicaAntiMatchExpression key not found, check for  rule:%v  mkey:%v mOperator:%v mValues:%v tval:%v", node, matchExp.Key, matchExp.Operator, matchExp.Values, tval)
+					switch matchExp.Operator {
+					case "DoesNotExist":
+						return false, &ErrFailedToInspectVolume{
+							Cause: fmt.Sprintf("ReplicaAntiAffinity with matchExpression Operator:'%v' is placed on node:%v not having label key:(%v:%v)", matchExp.Operator, node.Id, matchExp.Key, tval),
+						}
+					case "Exists":
+						logrus.Infof("ReplicaAntiMatchExpression  mkey:%v mOperator:%v does not exist on the node: %v", matchExp.Key, matchExp.Operator, node.Id)
+						return true, nil
+					default:
+						return false, &ErrFailedToInspectVolume{
+							Cause: fmt.Sprintf("ReplicaAntiAffinity with matchExpression Operator:'%v' is placed on node:%v not having label key:(%v:%v)", matchExp.Operator, node.Id, matchExp.Key, tval),
+						}
+					}
+
+				}
 			}
-		default:
 		}
 	}
 
-	logrus.Infof("Successfully inspected volume: %v (%v)", vol.Locator.Name, vol.Id)
-	return nil
+	return false, &ErrFailedToInspectVolume{
+		Cause: fmt.Sprintf("ReplicaAntiAffinity with matchExpression Operator:'%v' could not be validated for node:%v  having label key:(%v)", matchExp.Key, node.Id, matchExp.Key),
+	}
 }
 
+// VolumeMatchExpression
+func (d *portworx) VolumeMatchExpression(vol *api.Volume, matchExp *talisman_v1beta1.LabelSelectorRequirement, nodeReplList map[string][]*api.Volume, nodeGrp map[string]api.StorageNode) (bool, error) {
+	logrus.Infof("VolumeMatchExpression started nodeId: %v matchExpression:%v ", vol.Id, matchExp)
+
+	logrus.Debugf("VolumeMatchExpression rule:%v  mkey:%v mOperator:%v mValues:%v nodeReplList:%v, nodeGrp:%v", vol, matchExp.Key, matchExp.Operator, matchExp.Values, nodeReplList, nodeGrp)
+	if vol != nil {
+		// Check for Operators: In, Exists, NotIn, DoesNotExist
+		switch matchExp.Operator {
+		case "NotIn":
```

Comment:
lot of duplicate logic for variable operators. We need to have a common piece of code to create a list of current and expected values and then later on add a switch case based on operator to make the corresponding check.